File: scripts/enum/kms.js
'use strict';

const {
  KMSClient,
  ListKeysCommand,
  DescribeKeyCommand,
  GetKeyPolicyCommand,
  ListGrantsCommand,
  GetKeyRotationStatusCommand,
} = require('@aws-sdk/client-kms');

const { withRetry, paginate, createLogger } = require('../lib');
const { baseEnum } = require('../lib/base-enum');
const { extractPolicyPrincipals } = require('../lib/policy-parser');

// --- Helpers ---

function generateFindings(key) {
  const findings = [];

  if (!key.rotation_enabled) {
    findings.push({
      type: 'no_rotation',
      severity: 'medium',
      detail: 'Key rotation is not enabled',
    });
  }

  if (key.key_state === 'PendingDeletion') {
    findings.push({
      type: 'pending_deletion',
      severity: 'low',
      detail: 'Key is pending deletion',
    });
  }

  // Wildcard principal in policy
  if (key.policy_principals && key.policy_principals.includes('*')) {
    findings.push({
      type: 'wildcard_principal',
      severity: 'high',
      detail: 'Key policy allows wildcard principal',
    });
  }

  // Grants to external accounts
  if (key.grants && key.grants.length > 0) {
    findings.push({
      type: 'has_grants',
      severity: 'low',
      detail: `Key has ${key.grants.length} grant(s)`,
    });
  }

  return findings;
}

// --- Run (dependency-injectable) ---

async function run(opts = {}) {
  const runDir = opts.runDir;
  const region = opts.region;
  const accountId = opts.accountId;

  const client = opts.clients?.kms ?? new KMSClient({ region });

  const logger = opts.logger || createLogger(runDir, 'kms');
  logger.log('info', 'KMS_Enumeration_Start', { region });

  const findings = [];
  let status = 'complete';
  const errors = [];

  // ListKeys (paginated)
  let allKeys;
  try {
    logger.log('api_call', 'ListKeys', { service: 'kms' });
    allKeys = await paginate(client, ListKeysCommand, 'Keys', {});
  } catch (err) {
    logger.log('error', 'ListKeys', { error: err.message });
    await logger.flush();
    throw new Error(`ListKeys failed: ${err.message}`);
  }

  // Per-key: DescribeKey, filter to customer-managed
  for (const keyEntry of allKeys) {
    const keyId = keyEntry.KeyId;

    let keyMetadata;
    try {
      logger.log('api_call', 'DescribeKey', { key_id: keyId });
      const descResp = await withRetry(() =>
        client.send(new DescribeKeyCommand({ KeyId: keyId }))
      );
      keyMetadata = descResp.KeyMetadata;
    } catch (err) {
      errors.push({ resource: keyId, error: err.message });
      logger.log('warning', 'DescribeKey', { key_id: keyId, error: err.message });
      continue;
    }

    // Skip AWS-managed keys
    if (keyMetadata.KeyManager !== 'CUSTOMER') continue;

    const keyFinding = {
      resource_type: 'kms_key',
      resource_id: keyId,
      arn: keyMetadata.Arn,
      region,
      key_state: keyMetadata.KeyState,
      usage: keyMetadata.KeyUsage,
      origin: keyMetadata.Origin,
      description: keyMetadata.Description || '',
      rotation_enabled: false,
      policy_principals: [],
      grants: [],
      findings: [],
    };

    // GetKeyPolicy
    try {
      logger.log('api_call', 'GetKeyPolicy', { key_id: keyId });
      const policyResp = await withRetry(() =>
        client.send(new GetKeyPolicyCommand({ KeyId: keyId, PolicyName: 'default' }))
      );
      keyFinding.policy_principals = extractPolicyPrincipals(policyResp.Policy);
    } catch (err) {
      errors.push({ resource: keyId, error: `GetKeyPolicy: ${err.message}` });
      logger.log('warning', 'GetKeyPolicy', { key_id: keyId, error: err.message });
    }

    // ListGrants
    try {
      logger.log('api_call', 'ListGrants', { key_id: keyId });
      const grants = await paginate(client, ListGrantsCommand, 'Grants', {
        params: { KeyId: keyId },
      });
      keyFinding.grants = grants.map((g) => ({
        grant_id: g.GrantId,
        grantee_principal: g.GranteePrincipal,
        operations: g.Operations || [],
        retiring_principal: g.RetiringPrincipal || null,
      }));
    } catch (err) {
      errors.push({ resource: keyId, error: `ListGrants: ${err.message}` });
      logger.log('warning', 'ListGrants', { key_id: keyId, error: err.message });
    }

    // GetKeyRotationStatus
    try {
      logger.log('api_call', 'GetKeyRotationStatus', { key_id: keyId });
      const rotResp = await withRetry(() =>
        client.send(new GetKeyRotationStatusCommand({ KeyId: keyId }))
      );
      keyFinding.rotation_enabled = rotResp.KeyRotationEnabled || false;
    } catch (err) {
      // Some key types don't support rotation status
      const code = err.name || err.Code || '';
      if (code === 'UnsupportedOperationException') {
        keyFinding.rotation_enabled = null;
      } else {
        errors.push({ resource: keyId, error: `GetKeyRotationStatus: ${err.message}` });
        logger.log('warning', 'GetKeyRotationStatus', { key_id: keyId, error: err.message });
      }
    }

    keyFinding.findings = generateFindings(keyFinding);
    findings.push(keyFinding);
  }

  if (errors.length > 0) status = 'partial';

  await logger.flush();
  return { findings, status };
}

if (require.main === module) {
  baseEnum({ module: 'kms', run });
}

module.exports = { run };

File: scope/enumerators/kms.py
from __future__ import annotations

import json
from typing import Any

from scope.core.coverage import CoverageTracker
from scope.core.models import ModuleEnvelope


PRIMARY_CHECKS = ["list_keys"]
REQUIRED_CHECKS = ["describe_key", "key_policy", "grants"]


def run(factory: Any, region: str) -> ModuleEnvelope:
    tracker = CoverageTracker(primary=PRIMARY_CHECKS, required=REQUIRED_CHECKS)
    kms = factory.client("kms")
    findings: list[dict[str, Any]] = []

    try:
        keys = factory.paginate(kms, "list_keys", "Keys")
        tracker.record_ok("list_keys", resource="keys")
    except Exception as err:
        tracker.record_module_failure("list_keys", operation="kms.ListKeys", err=err)
        coverage, errors = tracker.to_envelope_fields()
        return ModuleEnvelope(
            module="kms",
            account_id=factory.account_id,
            region=region,
            status=tracker.derive_status(),
            resources=[],
            coverage=coverage,
            errors=errors,
        )

    for key_entry in keys:
        key_id = key_entry.get("KeyId")
        try:
            metadata = kms.describe_key(KeyId=key_id).get("KeyMetadata", {})
            tracker.record_ok("describe_key", resource=key_id)
        except Exception as err:
            tracker.record_resource_failure(
                "describe_key",
                resource=key_id,
                operation="kms.DescribeKey",
                err=err,
            )
            continue

        if metadata.get("KeyManager") != "CUSTOMER":
            continue

        key_arn = metadata.get("Arn")
        finding = {
            "resource_type": "kms_key",
            "resource_id": key_id,
            "arn": key_arn,
            "region": region,
            "key_state": metadata.get("KeyState"),
            "usage": metadata.get("KeyUsage"),
            "origin": metadata.get("Origin"),
            "description": metadata.get("Description") or "",
            "rotation_enabled": False,
            "rotation_status": None,
            "policy_principals": [],
            "policy_status": None,
            "grants": [],
            "grants_status": None,
            "findings": [],
        }

        try:
            response = kms.get_key_policy(KeyId=key_id, PolicyName="default")
            finding["policy_principals"] = _extract_policy_principals(response.get("Policy"))
            finding["policy_status"] = "present"
            tracker.record_ok("key_policy", resource=key_arn)
        except Exception as err:
            code = _error_code(err)
            finding["policy_status"] = _classify_status(code)
            tracker.record_resource_failure("key_policy", resource=key_arn, operation="kms.GetKeyPolicy", err=err)

        try:
            grants = factory.paginate(kms, "list_grants", "Grants", KeyId=key_id)
            finding["grants"] = [
                {
                    "grant_id": grant.get("GrantId"),
                    "grantee_principal": grant.get("GranteePrincipal"),
                    "operations": grant.get("Operations") or [],
                    "retiring_principal": grant.get("RetiringPrincipal"),
                }
                for grant in grants
            ]
            finding["grants_status"] = "present"
            tracker.record_ok("grants", resource=key_arn)
        except Exception as err:
            code = _error_code(err)
            finding["grants_status"] = _classify_status(code)
            tracker.record_resource_failure("grants", resource=key_arn, operation="kms.ListGrants", err=err)

        try:
            response = kms.get_key_rotation_status(KeyId=key_id)
            finding["rotation_enabled"] = response.get("KeyRotationEnabled") or False
            finding["rotation_status"] = "present"
            tracker.record_ok("rotation_status", resource=key_arn)
        except Exception as err:
            code = _error_code(err)
            if code == "UnsupportedOperationException":
                finding["rotation_enabled"] = None
                finding["rotation_status"] = "unsupported"
                tracker.record_skipped("rotation_status", resource=key_arn, reason="unsupported")
            else:
                finding["rotation_status"] = _classify_status(code)
                reason = "access_denied" if finding["rotation_status"] == "access_denied" else code
                tracker.record_skipped("rotation_status", resource=key_arn, reason=reason)

        finding["findings"] = _generate_findings(finding)
        findings.append(finding)

    coverage, errors = tracker.to_envelope_fields()
    return ModuleEnvelope(
        module="kms",
        account_id=factory.account_id,
        region=region,
        status=tracker.derive_status(),
        resources=findings,
        coverage=coverage,
        errors=errors,
    )


def _generate_findings(key: dict[str, Any]) -> list[dict[str, str]]:
    findings: list[dict[str, str]] = []
    if not key["rotation_enabled"]:
        findings.append(
            {
                "type": "no_rotation",
                "severity": "medium",
                "detail": "Key rotation is not enabled",
            }
        )
    if key["key_state"] == "PendingDeletion":
        findings.append(
            {
                "type": "pending_deletion",
                "severity": "low",
                "detail": "Key is pending deletion",
            }
        )
    if "*" in key.get("policy_principals", []):
        findings.append(
            {
                "type": "wildcard_principal",
                "severity": "high",
                "detail": "Key policy allows wildcard principal",
            }
        )
    grants = key.get("grants") or []
    if grants:
        findings.append(
            {
                "type": "has_grants",
                "severity": "low",
                "detail": f"Key has {len(grants)} grant(s)",
            }
        )
    return findings


def _extract_policy_principals(policy: str | dict[str, Any] | None) -> list[str]:
    if not policy:
        return []
    try:
        document = json.loads(policy) if isinstance(policy, str) else policy
    except json.JSONDecodeError:
        return []
    statements = document.get("Statement", []) if isinstance(document, dict) else []
    if isinstance(statements, dict):
        statements = [statements]

    principals: set[str] = set()
    for statement in statements:
        if not isinstance(statement, dict) or statement.get("Effect") != "Allow":
            continue
        principal = statement.get("Principal")
        if isinstance(principal, str):
            principals.add(principal)
        elif isinstance(principal, dict):
            for key in ("AWS", "Service", "Federated"):
                value = principal.get(key)
                if isinstance(value, str):
                    principals.add(value)
                elif isinstance(value, list):
                    principals.update(str(item) for item in value)
    return sorted(principals)


def _classify_status(code: str) -> str:
    return "access_denied" if code in {"AccessDenied", "AccessDeniedException"} else "error"


def _error_code(err: Exception) -> str:
    response = getattr(err, "response", None)
    if isinstance(response, dict):
        error = response.get("Error")
        if isinstance(error, dict) and error.get("Code"):
            return str(error["Code"])
    code = getattr(err, "Code", None)
    if code:
        return str(code)
    return err.__class__.__name__ or "Error"
